cdap-app-fabric/src/main/java/io/cdap/cdap/internal/tethering/TetheringServerHandler.java

```diff
@@ -222,4 +231,25 @@ private void updateTetherStatus(HttpResponder responder, String peer, TetheringS
     }
     responder.sendStatus(HttpResponseStatus.OK);
   }
+
+  /**
+   * Processes and publishes the list of tethering program updates received from the client
+   */
+  private void processProgramUpdateNotifications(FullHttpRequest request) throws BadRequestException {
+    List<Notification> notificationList;
+    try (Reader reader = new InputStreamReader(new ByteBufInputStream(request.content()), StandardCharsets.UTF_8)) {
+      notificationList = GSON.fromJson(reader, NOTIFICATION_LIST_TYPE);
+    } catch (JsonSyntaxException | IOException e) {
+      throw new BadRequestException("Unable to parse request: " + e.getMessage(), e);
+    }
+
+    try {
+      for (Notification notification : notificationList) {
+        messagingContext.getMessagePublisher().publish(NamespaceId.SYSTEM.getNamespace(), programStatusTopic,
+                                                       GSON.toJson(notification));
+      }
+    } catch (TopicNotFoundException | IOException e) {
+      LOG.error("Unable to publish program status update", e);
```

Comment:
should propagate these exceptions up so that the entire request fails. Otherwise the agent service can send a list of notifications, this fails to write them but returns a 200 so the agent service moves on and we lose data.

Reply:
fixed